rmc/src/service.rs

```diff
@@ -0,0 +1,420 @@
+/// for a high-level description of this protocol and how it is used for fork alerts.
+pub struct Service<H: Signable + Hash, MK: MultiKeychain> {
+    network_rx: UnboundedReceiver<RmcIncomingMessage<H, MK::Signature, MK::PartialMultisignature>>,
+    network_tx: UnboundedSender<RmcOutgoingMessage<H, MK>>,
+    scheduler: Box<dyn TaskScheduler<RmcHash<H, MK::Signature, MK::PartialMultisignature>>>,
+}
+
+impl<H: Signable + Hash + Eq + Clone + Debug, MK: MultiKeychain> Service<H, MK> {
+    pub fn new(
+        network_rx: UnboundedReceiver<
+            RmcIncomingMessage<H, MK::Signature, MK::PartialMultisignature>,
+        >,
+        network_tx: UnboundedSender<RmcOutgoingMessage<H, MK>>,
+        scheduler: impl TaskScheduler<RmcHash<H, MK::Signature, MK::PartialMultisignature>> + 'static,
+    ) -> Self {
+        Service {
+            network_rx,
+            network_tx,
+            scheduler: Box::new(scheduler),
+        }
+    }
+
+    fn handle_message(
+        &mut self,
+        message: RmcIncomingMessage<H, MK::Signature, MK::PartialMultisignature>,
+        handler: &mut Handler<H, MK>,
+    ) {
+        match message {
+            RmcIncomingMessage::StartRmc(hash) => {
+                debug!(target: "AlephBFT-rmc", "starting rmc for {:?}", hash);
+                let unchecked = handler.on_start_rmc(hash);
+                self.scheduler.add_task(RmcHash::SignedHash(unchecked));
+            }
+            RmcIncomingMessage::RmcHash(RmcHash::MultisignedHash(unchecked)) => {
+                match handler.on_multisigned_hash(&unchecked) {
+                    Ok(Some(multisigned)) => {
+                        self.scheduler.add_task(RmcHash::MultisignedHash(unchecked));
+                        if self
+                            .network_tx
+                            .unbounded_send(RmcOutgoingMessage::NewMultisigned(multisigned))
+                            .is_err()
+                        {
+                            error!(target: "AlephBFT-rmc", "network closed early.");
+                        }
+                    }
+                    Ok(None) => {}
+                    Err(error) => {
+                        warn!(target: "AlephBFT-rmc", "{}", error);
+                    }
+                }
+            }
+            RmcIncomingMessage::RmcHash(RmcHash::SignedHash(unchecked)) => {
+                match handler.on_signed_hash(&unchecked) {
+                    Ok(Some(multisigned)) => {
+                        self.scheduler.add_task(RmcHash::MultisignedHash(
+                            multisigned.clone().into_unchecked(),
+                        ));
+                        if self
+                            .network_tx
+                            .unbounded_send(RmcOutgoingMessage::NewMultisigned(multisigned))
+                            .is_err()
+                        {
+                            error!(target: "AlephBFT-rmc", "network closed early.");
```

Comment:
This just logs, below in `send_hash` it panics. Do you know why?

Reply:
yeah... quite inconsistent. I shall replace this code with `send_hash` and keep the panic as it was there previously